torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2247,88 +2385,220 @@ def _load_sharded_state_dict(
         with self.set_state_dict_type(StateDictType.SHARDED_STATE_DICT):
             return self.load_state_dict(state_dict, strict)
 
-    def forward(self, *args: Any, **kwargs: Any) -> Any:
+    def _reshard(
+        self,
+        params: List[FlatParameter],
+        free_full_params: bool,
+        free_mp_shard: bool,
+    ) -> None:
+        """
+        Reshards unsharded flattened parameters.
+
+        Args:
+            handles (List[FlatParamHandle]): Handles giving the flattened
+                parameters to reshard.
+            free_full_params (bool): Whether to free the unsharded flattened
+                parameter.
+            free_mp_shard (bool): Whether to free the reduced-precision sharded
+                flattened parameter.
+        """
+        if not params:
+            return
+        if free_full_params:
+            self._free_full_params(params)
+        if free_mp_shard:
+            self._free_mp_shard(params)
+        self._use_param_local_shard(params)
+
+    def _pre_forward_reshard(self, handles: List[FlatParamHandle]):
+        """
```

Comment:
This is just one option. Right now, the next module will reshard the parameters that were all-gathered in the previous module's `forward()`.

If we wanted to emulate the same behavior of the recursive wrapping, we can just move the reshard to the post-forward reshard.

Resharding in the pre-forward like this means that the last module's parameters (last with respect to `forward()` execution order) will not be resharded, which is exactly the effect we wanted by setting the root FSDP instance's `reshard_after_forward = False`.

It all depends on what you think is best for the scheduling algorithm. I have it like this so that we can reshard in the pre-forward, in the post-forward, or both. In other words, I am exposing two points for you to do this resharding.

Let me know if this is still confusing!

Reply:
If you can tell explicitly that a module will be the last module to execute in the forward, then I think the pre-forward reshard may be obsolete. However, if we cannot tell, then we may be able to use the pre-forward reshard and leverage the runtime execution to create that effect of `reshard_after_forward = False` for the last module.

We can always remove the pre-forward reshard if we do not find it useful.